### ISSUE TITLE
LoginRequest.Password is missing type parameter in json

### ISSUE THREAD

**Issue by jakobkmar** [Dominaezzz/matrix-kt#2]
Serializing LoginRequest.Password (and probably the other LoginRequest types as well) gives a wrong result, which is missing the type (e.g. `m.login.password`):
```kt
MatrixJson.encodeToString(LoginRequest.Password(UserIdentifier.Matrix("blue"), "mypassword"))
// output:
// {"identifier":{"type":"m.id.user","user":"blue"},"password":"mypassword"}
// what it should be:
// {"type":"m.login.password","identifier":{"type":"m.id.user","user":"blue"},"password":"mypassword"}
```

This (of course) leads to the following exception when trying to call the `login` function:
```
Exception in thread "main" Unknown(error=Missing JSON keys.)
```

**Comment by Dominaezzz** [comment]
Ah, tricky.
You need to do this.
```
MatrixJson.encodeToString<LoginRequest>(LoginRequest.Password(UserIdentifier.Matrix("blue"), "mypassword"))
```

You have to serialize using the base class to get the `type`, if you use the subclass `kotlinx.serialization` will miss out the discriminator. It's a little weird but it'll make sense after a while.

Are you using `MatrixClient` or a plain `HttpClient` (without rpc), my apis should make this "just work".

**Comment by Dominaezzz** [comment]
Oh no, wait, I just looked at the code again and realised it won't work. Rip.
That's annoying, I know have to figure out how to tell ktor to do the right thing.
Thanks for making an issue for this.

**Comment by Dominaezzz** [comment]
Here's a workaround that you can use.
```
val PatchedJson = Json(MatrixJson) {
    serializersModule += SerializersModule {
        contextual(LoginRequest.Password::class, LoginRequest.serializer() as KSerializer<LoginRequest.Password>)
    }
}
```

Then passing `PatchedJson` into the `MatrixClient` builder like this should work.
```
MatrixClient(...) {
    Json { serializer = KotlinxSerializer(PatchedJson) }
}
```

**Comment by bluefireoly** [comment]
Thank you, I will try this out, do you spontaneously know any other places where this problem appears (or is it really just a problem with LoginRequest)?

**Comment by Dominaezzz** [comment]
Just `LoginRequest`, it's the only one with a polymorphic body.

**Comment by bluefireoly** [comment]
The workaround fixed it for now, on `0.1.4` I just passed the `PatchedJson` instance as a parameter to the `MatrixConfig` function.

**Comment by Dominaezzz** [comment]
https://youtrack.jetbrains.com/issue/KTOR-744

**Comment by Dominaezzz** [comment]
Fixed in 0.1.9
